File: Par_Algo_Direct/llm/qwen/python/lattice.py
#!/usr/bin/env python3
"""
Formal Concept Lattice Generator (NextClosure Algorithm)
--------------------------------------------------------
Lit un contexte formel binaire (CSV), calcule le treillis de concepts formels,
exporte le diagramme de Hasse au format Graphviz (.dot) et mesure le temps CPU.
Conforme aux spécifications FCA, PEP 8, et contraintes d'exécution imposées.
"""

import sys
import csv
import time
import os
from typing import List, Set, Tuple, Dict, Optional


def read_and_validate_csv(filepath: str) -> Tuple[List[str], List[str], List[List[int]]]:
    """
    Lit et valide le fichier CSV.
    Retourne: (noms_objets, noms_attributs, matrice_contexte)
    Lève ValueError si le format ou les données sont invalides.
    """
    if not os.path.isfile(filepath):
        raise FileNotFoundError(f"Le fichier '{filepath}' est introuvable.")

    objects = []
    attributes = []
    context = []

    with open(filepath, "r", encoding="utf-8-sig") as f:
        reader = csv.reader(f, delimiter=";")
        
        # Lecture de l'en-tête
        try:
            header = next(reader)
        except StopIteration:
            raise ValueError("Fichier CSV vide.")
            
        # La première cellule est souvent vide ou contient un placeholder. On ignore.
        attributes = [a.strip() for a in header[1:] if a.strip()]
        if not attributes:
            raise ValueError("Aucun attribut trouvé dans l'en-tête.")

        num_attrs = len(attributes)

        for line_num, row in enumerate(reader, start=2):
            if not row or all(c.strip() == "" for c in row):
                continue  # Ligne vide ignorée
                
            obj_name = row[0].strip()
            if not obj_name:
                raise ValueError(f"Ligne {line_num}: nom d'objet manquant.")
            objects.append(obj_name)

            # Extraction et validation des valeurs binaires
            values = []
            raw_vals = row[1:]
            if len(raw_vals) < num_attrs:
                raise ValueError(f"Ligne {line_num}: nombre d'attributs insuffisant.")
                
            for i, val in enumerate(raw_vals[:num_attrs]):
                val = val.strip()
                if val not in ("0", "1"):
                    raise ValueError(f"Ligne {line_num}, attribut {i}: valeur '{val}' invalide (attendu 0 ou 1).")
                values.append(int(val))
            context.append(values)

    if not objects:
        raise ValueError("Aucun objet trouvé dans le contexte.")

    return objects, attributes, context


def precompute_mappings(context: List[List[int]], num_attrs: int) -> List[Set[int]]:
    """
    Précalcule pour chaque attribut l'ensemble des objets qui le possèdent.
    Optimise les opérations de dérivation ultérieures.
    """
    num_objs = len(context)
    attr_to_objs: List[Set[int]] = [set() for _ in range(num_attrs)]
    for obj_idx, row in enumerate(context):
        for attr_idx, val in enumerate(row):
            if val == 1:
                attr_to_objs[attr_idx].add(obj_idx)
    return attr_to_objs


def derive_and_close(
    attr_indices: Set[int],
    attr_to_objs: List[Set[int]],
    all_obj_indices: Set[int],
    all_attr_indices: Set[int]
) -> Tuple[Set[int], Set[int]]:
    """
    Calcul de la fermeture (intention, extension) d'un ensemble d'attributs.
    Étape mathématique clé : 
      1. Dérivation vers les objets : A' = ⋂_{m∈A} m'
      2. Dérivation inverse vers les attributs : A'' = {m ∈ M | A' ⊆ m'}
    Retourne (intention_fermée, extension_fermée)
    """
    # 1. Calcul de l'extension (intersection des objets possédant tous les attributs)
    if not attr_indices:
        extent = all_obj_indices.copy()
    else:
        it = iter(attr_indices)
        extent = attr_to_objs[next(it)].copy()
        for a in it:
            extent.intersection_update(attr_to_objs[a])
            if not extent:
                break

    # 2. Calcul de l'intention fermée depuis l'extension
    intent = set()
    for m in all_attr_indices:
        if extent.issubset(attr_to_objs[m]):
            intent.add(m)
    return intent, extent


def generate_lattice(
    num_attrs: int,
    attr_to_objs: List[Set[int]],
    all_obj_indices: Set[int],
    all_attr_indices: Set[int]
) -> List[Tuple[Set[int], Set[int]]]:
    """
    Algorithme NextClosure (Ganter, 1984) : génération itérative de tous les concepts
    dans l'ordre lectic, sans récursion ni recalcul superflu.
    """
    concepts: List[Tuple[Set[int], Set[int]]] = []
    
    # Concept initial : fermeture de l'ensemble vide (borne inférieure)
    current_intent, current_extent = derive_and_close(
        set(), attr_to_objs, all_obj_indices, all_attr_indices
    )
    concepts.append((current_intent, current_extent))

    while True:
        next_concept = None
        
        # Parcours des attributs du plus grand indice au plus petit (ordre lectic descendant)
        for m in range(num_attrs - 1, -1, -1):
            if m in current_intent:
                continue  # L'attribut est déjà dans l'intention courante

            # A = B ∩ {0, ..., m-1}
            lower_bound_attrs = set(range(m))
            A = current_intent & lower_bound_attrs

            # Calcul de la fermeture candidate C = (A ∪ {m})''
            cand_intent, cand_extent = derive_and_close(
                A | {m}, attr_to_objs, all_obj_indices, all_attr_indices
            )

            # Condition de NextClosure :
            # 1. m doit appartenir à l'intention candidate (garantit l'ajout de m)
            # 2. La partie inférieure de C doit être identique à celle de B
            #    (garantit l'ordre lectic strict sans saut)
            if m in cand_intent and (cand_intent & lower_bound_attrs) == (current_intent & lower_bound_attrs):
                next_concept = (cand_intent, cand_extent)
                break  # Premier m valide trouvé = successeur lectic immédiat

        if next_concept is None:
            break  # Aucun successeur : fin du treillis (borne supérieure atteinte)
            
        concepts.append(next_concept)
        current_intent, current_extent = next_concept

    return concepts


def build_hasse_diagram(concepts: List[Tuple[Set[int], Set[int]]]) -> List[Tuple[int, int]]:
    """
    Calcule les relations de couverture (Hasse) du treillis.
    Un concept j couvre i si Ext(i) ⊂ Ext(j) et ∄k tel que Ext(i) ⊂ Ext(k) ⊂ Ext(j).
    """
    n = len(concepts)
    edges: List[Tuple[int, int]] = []  # (enfant_id, parent_id)
    
    # Pré-calcul des inclusions strictes pour optimiser
    is_strict_subset = [[False] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if i != j and concepts[i][1].issubset(concepts[j][1]) and concepts[i][1] != concepts[j][1]:
                is_strict_subset[i][j] = True

    for i in range(n):
        for j in range(n):
            if not is_strict_subset[i][j]:
                continue
                
            # Vérifier s'il existe un intermédiaire k
            is_cover = True
            for k in range(n):
                if is_strict_subset[i][k] and is_strict_subset[k][j]:
                    is_cover = False
                    break
            if is_cover:
                edges.append((i, j))
                
    return edges


def generate_dot(
    concepts: List[Tuple[Set[int], Set[int]]],
    edges: List[Tuple[int, int]],
    obj_names: List[str],
    attr_names: List[str]
) -> str:
    """
    Génère la chaîne DOT conforme au format imposé.
    Applique le styling conditionnel aux nœuds extrémaux et pivot.
    """
    n = len(concepts)
    num_attrs = len(attr_names)
    
    # Identification des concepts extrémaux et pivot
    bottom_idx = 0  # NextClosure commence toujours par l'intention vide
    top_idx = next(i for i, (intent, extent) in enumerate(concepts) if not extent)
    
    # Pivot : premier concept intermédiaire (ni borne inf, ni borne sup)
    pivot_idx = next((i for i in range(n) if i != bottom_idx and i != top_idx), None)

    lines = ["digraph G {", "    rankdir=BT;"]

    for idx, (intent, extent) in enumerate(concepts):
        # Formatage des listes
        attr_list = "\n".join(attr_names[a] for a in sorted(intent)) if intent else ""
        obj_list = "\n".join(obj_names[o] for o in sorted(extent)) if extent else ""
        
        # Construction du label record
        label = f"{{{idx} (I: {len(intent)}, E: {len(extent)})|{attr_list}|{obj_list}}}"
        
        # Application du styling conditionnel
        if idx == bottom_idx or idx == top_idx:
            node_def = f'    {idx} [shape=record,style=filled,fillcolor=lightblue,label="{label}"];'
        elif idx == pivot_idx:
            node_def = f'    {idx} [shape=record,style=filled,fillcolor=orange,label="{label}"];'
        else:
            node_def = f'    {idx} [shape=record,label="{label}"];'
            
        lines.append(node_def)

    for child, parent in edges:
        lines.append(f"    {child} -> {parent}")

    lines.append("}")
    return "\n".join(lines)


def main() -> None:
    """Point d'entrée principal : orchestration, mesure CPU et I/O."""
    if len(sys.argv) != 2:
        print("Usage: python lattice_generator.py <chemin_vers_csv>")
        sys.exit(1)

    csv_path = sys.argv[1]

    # 1. Lecture et validation
    try:
        obj_names, attr_names, context = read_and_validate_csv(csv_path)
    except Exception as e:
        print(f"Erreur de validation CSV : {e}", file=sys.stderr)
        sys.exit(1)

    num_attrs = len(attr_names)
    all_attr_idx = set(range(num_attrs))
    all_obj_idx = set(range(len(obj_names)))
    attr_to_objs = precompute_mappings(context, num_attrs)

    # 2. Calcul du treillis (mesure CPU stricte)
    start_cpu = time.process_time()
    concepts = generate_lattice(num_attrs, attr_to_objs, all_obj_idx, all_attr_idx)
    edges = build_hasse_diagram(concepts)
    end_cpu = time.process_time()

    # 3. Génération et écriture DOT
    dot_content = generate_dot(concepts, edges, obj_names, attr_names)
    output_path = os.path.join(os.path.dirname(os.path.abspath(csv_path)), "output.dot")
    with open(output_path, "w", encoding="utf-8") as f:
        f.write(dot_content)

    # 4. Affichage du temps CPU
    print(f"Temps CPU : {end_cpu - start_cpu:.4f} secondes")
    print(f"Treillis généré : {len(concepts)} concepts, {len(edges)} arêtes.")
    print(f"Exporté vers : {output_path}")


if __name__ == "__main__":
    main()

File: Par_Algo_Direct/llm/qwen/c++/lattice.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>
#include <chrono>
#include <cmath>

using namespace std;

// Structure représentant un concept formel (intention/extension)
struct Concept {
    vector<bool> intent;
    vector<bool> extent;
};

// Vérifie si a est un sous-ensemble strict de b
inline bool is_proper_subset(const vector<bool>& a, const vector<bool>& b) {
    bool strict = false;
    for (size_t i = 0; i < a.size(); ++i) {
        if (a[i] && !b[i]) return false;
        if (b[i] && !a[i]) strict = true;
    }
    return strict;
}

// Compare lexicographiquement : retourne vrai si a > b
inline bool is_lexicographically_greater(const vector<bool>& a, const vector<bool>& b) {
    for (size_t i = 0; i < a.size(); ++i) {
        if (a[i] != b[i]) return a[i] > b[i];
    }
    return false;
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        cerr << "Usage: " << argv[0] << " input.csv output.dot\n";
        return 1;
    }

    ifstream fin(argv[1]);
    if (!fin) { cerr << "Erreur: impossible d'ouvrir " << argv[1] << "\n"; return 1; }

    // 1. Parsing CSV
    vector<string> attr_names;
    string line;
    if (getline(fin, line)) {
        stringstream ss(line);
        string token;
        while (getline(ss, token, ';')) {
            if (!token.empty()) {
                if (token.back() == '\r') token.pop_back();
                attr_names.push_back(token);
            }
        }
    }
    size_t num_attrs = attr_names.size();
    if (num_attrs == 0) { cerr << "Erreur: aucun attribut trouvé\n"; return 1; }

    vector<string> obj_names;
    vector<vector<bool>> attr_extents(num_attrs);

    while (getline(fin, line)) {
        if (line.empty() || line.find_first_not_of(" \t\r\n") == string::npos) continue;
        stringstream ss(line);
        string obj_name;
        getline(ss, obj_name, ';');
        if (!obj_name.empty() && obj_name.back() == '\r') obj_name.pop_back();
        obj_names.push_back(obj_name);

        size_t obj_idx = obj_names.size() - 1;
        for (size_t j = 0; j < num_attrs; ++j) {
            string val;
            getline(ss, val, ';');
            if (!val.empty() && val.back() == '\r') val.pop_back();
            if (j >= attr_extents.size()) attr_extents.resize(j + 1);
            attr_extents[j].push_back(val == "1");
        }
    }
    size_t num_objs = obj_names.size();

    // Fonction de fermeture : closure(X) = (X')'
    auto closure = [&](const vector<bool>& intent) -> vector<bool> {
        vector<bool> extent(num_objs, true);
        for (size_t a = 0; a < num_attrs; ++a) {
            if (intent[a]) {
                for (size_t o = 0; o < num_objs; ++o) {
                    if (!attr_extents[a][o]) extent[o] = false;
                }
            }
        }
        vector<bool> new_intent(num_attrs, false);
        for (size_t a = 0; a < num_attrs; ++a) {
            bool shared = true;
            for (size_t o = 0; o < num_objs; ++o) {
                if (extent[o] && !attr_extents[a][o]) { shared = false; break; }
            }
            new_intent[a] = shared;
        }
        return new_intent;
    };

    // Calcul d'extension simple (sans recomputer l'intention)
    auto compute_extent = [&](const vector<bool>& intent) -> vector<bool> {
        vector<bool> extent(num_objs, true);
        for (size_t a = 0; a < num_attrs; ++a) {
            if (intent[a]) {
                for (size_t o = 0; o < num_objs; ++o) {
                    if (!attr_extents[a][o]) extent[o] = false;
                }
            }
        }
        return extent;
    };

    // 2. Calcul du treillis via NextClosure (Ganter)
    // Algorithme "one-pass" : parcours itératif de l'espace lexicographique des intentions fermées.
    // Principe : À partir de l'intention fermée courante A, on cherche le plus grand indice i tel que m_i ∉ A.
    // On construit B = (A ∩ {m < m_i}) ∪ {m_i}, on ferme B. Si closure(B) >_lex A, c'est le concept suivant.
    auto start_time = chrono::high_resolution_clock::now();

    vector<Concept> concepts;
    vector<bool> A(num_attrs, false);
    A = closure(A);
    concepts.push_back({A, compute_extent(A)});

    while (true) {
        bool found_next = false;
        for (int i = (int)num_attrs - 1; i >= 0; --i) {
            if (!A[i]) {
                vector<bool> B = A;
                for (int j = i + 1; j < (int)num_attrs; ++j) B[j] = false;
                B[i] = true;
                vector<bool> C = closure(B);
                if (is_lexicographically_greater(C, A)) {
                    A = C;
                    concepts.push_back({A, compute_extent(A)});
                    found_next = true;
                    break;
                }
            }
        }
        if (!found_next) break;
    }

    // 3. Calcul de la relation de couverture
    vector<pair<int, int>> edges;
    size_t C = concepts.size();
    for (size_t j = 0; j < C; ++j) {
        vector<int> parents;
        for (size_t i = 0; i < C; ++i) {
            if (i == j) continue;
            if (is_proper_subset(concepts[i].extent, concepts[j].extent))
                parents.push_back((int)i);
        }
        for (size_t p = 0; p < parents.size(); ++p) {
            bool is_cover = true;
            for (size_t q = 0; q < parents.size(); ++q) {
                if (p == q) continue;
                if (is_proper_subset(concepts[parents[p]].extent, concepts[parents[q]].extent)) {
                    is_cover = false; break;
                }
            }
            if (is_cover) edges.emplace_back((int)j, parents[p]);
        }
    }

    auto end_time = chrono::high_resolution_clock::now();
    long long ms = llround(chrono::duration<double, milli>(end_time - start_time).count());
    cout << "Temps CPU calcul treillis : " << ms << " ms\n";

    // 4. Export DOT strict
    ofstream fout(argv[2]);
    if (!fout) { cerr << "Erreur: impossible d'écrire " << argv[2] << "\n"; return 1; }

    fout << "digraph G { rankdir=BT;\n";
    for (size_t id = 0; id < C; ++id) {
        int I_cnt = 0; for (bool b : concepts[id].intent) if (b) I_cnt++;
        int E_cnt = 0; for (bool b : concepts[id].extent) if (b) E_cnt++;

        string color_attr = "";
        if (I_cnt == 0 || E_cnt == 0) color_attr = ",fillcolor=lightblue";
        else if (I_cnt >= 5 && E_cnt >= 3) color_attr = ",fillcolor=orange";

        string intent_str, extent_str;
        bool first = true;
        for (size_t a = 0; a < num_attrs; ++a) {
            if (concepts[id].intent[a]) {
                if (!first) intent_str += "\n";
                intent_str += attr_names[a];
                first = false;
            }
        }
        first = true;
        for (size_t o = 0; o < num_objs; ++o) {
            if (concepts[id].extent[o]) {
                if (!first) extent_str += "\n";
                extent_str += obj_names[o];
                first = false;
            }
        }

        string label = "{" + to_string(id) + " (I: " + to_string(I_cnt) + ", E: " + to_string(E_cnt) + ")|";
        if (!intent_str.empty()) label += intent_str + "\n";
        label += "|";
        if (!extent_str.empty()) label += extent_str + "\n";
        label += "}";

        fout << "    " << id << " [shape=record,style=filled" << color_attr << ",label=\"" << label << "\"];\n";
    }

    for (const auto& e : edges) {
        fout << "    " << e.first << " -> " << e.second << "\n";
    }
    fout << "}\n";

    return 0;
}